handler/oauth2/flow_authorize_code_token.go

```diff
@@ -198,7 +196,7 @@ func (c *AuthorizeExplicitGrantHandler) PopulateTokenEndpointResponse(ctx contex
 		responder.SetExtra("refresh_token", refresh)
 	}
 
-	if err := storage.MaybeCommitTx(ctx, c.CoreStorage); err != nil {
+	if err = storage.MaybeCommitTx(ctx, c.CoreStorage); err != nil {
```

Comment:
One problem I have observed with this control flow is that RDBMS roll back transactions automatically when commits fail. When rolling back a failed transaction, we will always end up with an error because there is nothing to roll back any more. This in turn causes the root of the issue (e.g. duplicate key) to be quite difficult to find/debug, because it is hidden in the rollback error and is probably missing content.

We had this issue in Hydra, the fix is to just return commit errors normally.

Reply:
> One problem I have observed with this control flow is that RDBMS roll back transactions automatically when commits fail.

Yes, I note this (and rationale for this change) in the description of the PR:

> There is one logical change I have done: before rollback was not called if commit failed, now I made it that it is called if commit fails. It is generally safe to call rollback outside of a transaction in PostgreSQL (you only get warning), but if the connection is in error state, then rollback is required. So just to be safe, I think it is better to rollback on any error, including when commit fails (because we might not know what is the reason for failure of that call and in what state is the connection).

> When rolling back a failed transaction, we will always end up with an error because there is nothing to roll back any more.

Hm, at least in [PostgreSQL you do not get an error](https://www.postgresql.org/docs/devel/sql-rollback.html):

> Issuing `ROLLBACK` outside of a transaction block emits a warning and otherwise has no effect.

With which database have you noticed the error?

> This in turn causes the root of the issue (e.g. duplicate key) to be quite difficult to find/debug, because it is hidden in the rollback error and is probably missing context.

I think this can be fixed a bit. So regular error looks like:

`errorsx.WithStack(fosite.ErrServerError.WithWrap(storageErr).WithDebug(storageErr.Error()))`

Current rollback error hides the regular error completely:

`err = errorsx.WithStack(fosite.ErrServerError.WithWrap(rbErr).WithDebug(rbErr.Error()))`

But in `PopulateTokenEndpointResponse` we have in fact a precedence which I think addresses the issue:

`errorsx.WithStack(fosite.ErrServerError.WithWrap(err).WithDebugf("error: %s; rollback error: %s", err, rollBackTxnErr))`

I think I could adopt this for rollback error everywhere. So:

`errorsx.WithStack(fosite.ErrServerError.WithWrap(storageErr).WithDebugf("error: %s; rollback error: %s", storageErr, rbErr))`

Reply:
> With which database have you noticed the error?

I don't quite remember, I think it was either MySQL or CockroachDB :/

> I think I could adopt this for rollback error everywhere. So:

SGTM! :)

Reply:
Done.